### DISCUSSION TITLE
Support for server side rendering

### DISCUSSION THREAD

**Discussion by PKASHYAP-BEC** [iTwin/itwinjs-core#184] [Show and tell] [answered]
Hello, 
We are trying to use imodeljs libraries in Server side rendering and getting a lot of issues. 
Can someone confirm,if imodeljs supports SSR or have any plan for enabling SSR in near future?

Thank you

**Comment by pmconne** [comment]
Please elaborate on "lots of issues" and provide details of your server environment.
The iModel.js frontend is designed to run in a browser or electron environment. Graphics capabilities are often constrained or unavailable in server contexts.

**Reply by hsanand61564** [reply]
@pmconne The basic issue that we are facing is when we try server-side rendering, global objects like a window, document are undefined so we are getting errors like 

ReferenceError: document is not defined
    at \node_modules\@bentley\imodeljs-frontend\lib\IModeljs-css.js:111:19

do have any workaround for these kinds of issues

**Comment by pmconne** [comment]
`window`, `document`, etc are supplied by the browser (or electron); outside of that context they will be undefined.
Short of running a browser on your server, there isn't really a workaround.
Can I ask what you're trying to achieve by running the frontend on a server?

**Reply by PKASHYAP-BEC** [reply]
@pmconne we are trying to use our react based imodel viewer as a micro frontend using SSR. 

The rest of our micro frontends are running great using SSR than client rendering, so we are trying to use SSR for our viewer also.

**Comment by pmconne** [comment]
Thanks. I don't think SSR can be made to work for an interactive iModel viewer.

**Comment by wgoehrig** [comment, marked as answer]
Just want to clarify, since I think there's been a bit of a miscommunication here regarding what kind of SSR we're talking about.

@pmconne is right that rendering viewports will always require an actual browser, but assuming you mean **[React Server Side Rendering](https://reactjs.org/docs/react-dom-server.html)** - that should (at least in theory) be partially supported.  We are working on fixing those `document is not defined` errors you're getting now.

However, I should also note that we are not doing any SSR of our frontends and have no plans to at this time, so there's a good chance it takes a bit more work to get SSR fully setup.  We assume all iModel.js frontends are largely interactive single-page applications, so I wouldn't expect to see much immediate benefit to enabling SSR.  In short, we've found SSR to be unnecessary, but YMMV.

**Reply by calebmshafer** [reply]
Here is the fix that @wgoehrig mentions, #295, to follow the progress of this being integrated.  Once that is merged into `master` it will be available in the next nightly release. (if merged today will be `2.10.0-dev.14`).

**Reply by calebmshafer** [reply]
@PKASHYAP-BEC this fix has been published in 2.10.0-dev.15. Let us know if you hit additional issues with Server Side Rendering and create a GitHub Issue for it. Thanks!

**Reply by hsanand61564** [reply]
@calebmshafer We tried with version 2.10.0 and the issue for this still persists there. Have you guys released the solution for Server Side Rendering in the version?

It is still not able to find the document object

**Reply by calebmshafer** [reply]
@hsanand61564 The fix I made was in 2.10.0. Would you mind providing instructions on how you're setting up SSR or a repo with it reproducible?  The testing I did must not have covered all cases.

**Reply by hsanand61564** [reply]
@calebmshafer I will try giving you the idea as to how we are trying to implement the same.

1. In the below code we are making the viewer
![image](https://user-images.githubusercontent.com/60655118/103002634-41002f80-4555-11eb-95f8-52ee114d358a.png)

we are using the above viewer as a package in our MFE(Micro Frontends)

2. In the code below the App component is the viewer from 1st step
![image](https://user-images.githubusercontent.com/60655118/103002721-6bea8380-4555-11eb-953b-74202a2e86a7.png)

3.  Below snippet is having component DataGrid which is the outcome from Step 2

![image](https://user-images.githubusercontent.com/60655118/103003628-1c0cbc00-4557-11eb-8995-06c347cf5a6f.png)

4. Finally we are doing the below thing
![image](https://user-images.githubusercontent.com/60655118/103003039-0d71d500-4556-11eb-832a-bccc8261e6ca.png)

![image](https://user-images.githubusercontent.com/60655118/103003004-00ed7c80-4556-11eb-88d1-dd1c96e16092.png)

After doing all the above steps we build the solution and then try to run the same and getting the following error

![image](https://user-images.githubusercontent.com/60655118/103003228-6a6d8b00-4556-11eb-868c-393ed35a121d.png)


Hope the above details help us in some manner

**Reply by hsanand61564** [reply]
@calebmshafer any updates?

**Reply by calebmshafer** [reply]
@hsanand61564  just had a chance to take a look and judging by the error messages, there is most likely other assumptions that `window` exists in layers above the imodeljs-frontend package.  This is most likely in the ui-* packages. I'm taking a pass through right now to identify them and open a pull request with a fix.

**Comment by hsanand61564** [comment]
@calebmshafer @pmconne Any updates on the reply

**Reply by calebmshafer** [reply]
@hsanand61564 just got back from Holiday today.  I'll take a look at your response and provide an update later today.

**Reply by hsanand61564** [reply]
@calebmshafer Any updates?

**Reply by hsanand61564** [reply]
@pmconne @calebmshafer Any Updates?

**Reply by hsanand61564** [reply]
@calebmshafer @pmconne Any updates on the same?

**Reply by pmconne** [reply]
@calebmshafer was going to look through the UI packages for usage of browser-specific APIs like `window` and `document`. A quick grep through their code reveals many usages of `window`. This is unsurprising since the UI packages naturally assume they are executing in a browser context. I think we'll need input from @NancyMcCallB regarding whether and how the UI packages could be used in a server-side context.

**Comment by NancyMcCallB** [comment]
Hi, @hsanand61564,

The AppUi system for iTwin.js has some features and implementation details that make it unfriendly to SSR, including:
•	Window event monitoring
•	Dynamic calculation of overflow for containers like the Tool Settings and the Status Bar
•	Save/restore for the user’s frontstage layout
 
Refactoring for AppUi for SSR might be possible, but not for a minor release. It would be helpful if you could tell us what you want to do with SSR and what parts of the UI you’re using.
